File: agent/commands-generator/src/main/java/org/matilda/commands/python/writer/PythonFileWriter.java
package org.matilda.commands.python.writer;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;

public class PythonFileWriter {
    private final File mBaseFile;

    public PythonFileWriter(File baseFile) {
        mBaseFile = baseFile;
    }

    public void write(PythonFile pythonFile) throws IOException {
        File file = getFile(pythonFile);
        ensureParentDirectoriesExist(file);
        try (FileWriter fileWriter = new FileWriter(file)) {
            fileWriter.write(pythonFile.getContent());
        }
    }

    private File getFile(PythonFile pythonFile) {
        return new File(mBaseFile, pythonFile.getPackage().toPath() + ".py");
    }

    private void ensureParentDirectoriesExist(File file) throws IOException {
        File parentFile = file.getParentFile();
        if (!parentFile.exists()) {
            if (!parentFile.mkdirs()) {
                throw new IOException("Failed to ensure that directory exists: " + parentFile);
            }
        }
    }
}

File: agent/commands-generator/src/main/java/org/matilda/commands/python/writer/PythonFileWriter.kt
package org.matilda.commands.python.writer

import java.io.File
import java.io.IOException

class PythonFileWriter(private val mBaseFile: File) {
    fun write(pythonFile: PythonFile) {
        val file = getFile(pythonFile)
        ensureParentDirectoriesExist(file)
        file.writeText(pythonFile.content)
    }

    private fun getFile(pythonFile: PythonFile) = File(mBaseFile, pythonFile.packageName.toPath() + ".py")

    private fun ensureParentDirectoriesExist(file: File) {
        val parentFile = file.parentFile
        if (!parentFile.exists()) {
            if (!parentFile.mkdirs()) {
                throw IOException("Failed to ensure that directory exists: $parentFile")
            }
        }

        generateInitFiles(file)
    }

    private fun generateInitFiles(file: File) {
        var parentDir = file.parentFile
        while (parentDir != mBaseFile) {
            val initFile = File(parentDir, "__init__.py")
            initFile.createNewFile()
            parentDir = parentDir.parentFile
        }
    }
}
